### PR TITLE
Fix leak in dwarf processing.

### PR DESCRIPTION
**Pull request by karliss**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

Store read files with line information in in a dictionary. Simplifies the logic within `bin_dwarf` and prevents memory leak. This also reduces execution time by not having to process the same file multiple times. Previous version only stored last 1-2 files. 

As usual there is room for further optimizations but it isn't the worst part anymore. There is little value for doing it right now before improve other parts first.

When testing on Cutter debug executable
* prevent 6.7 GB memory leak
* reduce peak memory usage from 6.7GB to 1.9GB
* reduce execution time 49s -> 20s
memory consumption screenshot

**Test plan**

* comparisons performed by running `r2 -ci -qq executable`

Wait for coverage report to check if modifed code is executed during tests, based on that decide need for more tests.

**Closing issues**

### CODE REVIEW

File: libr/core/cbin.c [hunk 1]
```diff
@@ -1015,6 +1015,41 @@ static int bin_info(RCore *r, int mode, ut64 laddr) {
...
+static void file_lines_free(FileLines *file) {
+	if (!file) {
+		return;
+	}
+	R_FREE (file->line_starts);
+	R_FREE (file->content);
+	R_FREE (file);
+}
+
+FileLines *read_file_lines (const char *path) {
```
Comment (ret2libc) [inline libr/core/cbin.c:1033]:
```suggestion
FileLines *read_file_lines(const char *path) {
```

**Review by ret2libc — COMMENTED** [review]
LGTM

**Comment by karliss** [comment]
The reported test coverage looks good, but it seems that none of them use source code lines read from file. I tried replacing reading of file with returning fixed string and nothing failed. I will try to investigate it a bit more to see if test detecting it can be easily added.

File: libr/core/cbin.c [hunk 2]
```diff
@@ -1015,6 +1015,41 @@ static int bin_info(RCore *r, int mode, ut64 laddr) {
...
+	int *line_starts;
+	char *content;
+	int line_count;
+} FileLines;
+
+static void file_lines_free(FileLines *file) {
+	if (!file) {
+		return;
+	}
+	R_FREE (file->line_starts);
```
Comment (trufae) [inline libr/core/cbin.c:1028]: just free, without nullifying the pointers

File: libr/core/cbin.c [hunk 3]
```diff
@@ -1015,6 +1015,41 @@ static int bin_info(RCore *r, int mode, ut64 laddr) {
 	return true;
 }
 
+typedef struct {
+	int *line_starts;
```
Comment (trufae) [inline libr/core/cbin.c:1019]: size_t

File: libr/core/cbin.c [hunk 4]
```diff
@@ -1015,6 +1015,41 @@ static int bin_info(RCore *r, int mode, ut64 laddr) {
...
+	char *content;
+	int line_count;
```
Comment (trufae) [inline libr/core/cbin.c:1021]: size_t

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cbin.c (hunk 3 above)
Comment (karliss) [reply libr/core/cbin.c:1019]: That also means changing the signature of r_str_split_lines and all the code using it. I can do it if you want, there aren't too many such places.

Comment (trufae) [reply libr/core/cbin.c:1019]: Yes i understand, just trying to be more formal and correct on every PR we merge. Not sure what @ret2libc thinks about that

**Comment by karliss** [comment]
Looked into testing. It's somewhat problematic because dwarf test executables have absolute paths to the environement in which they were compiled. And options for source file prefix and substitution are lacking. Manually verified using `idj~{}` that it reads source lines for locally compiled executable.

Although codecov reported that lines are executed. So maybe there is an appropriate executable in testsuite with correct relative paths.

File: libr/core/cbin.c (hunk 3 above)
Comment (karliss) [reply libr/core/cbin.c]: Refactored in separate commit.

**Review by trufae — APPROVED** [review]

**Pull request merged**